### PR TITLE
perf(allocator): lazy-batch free() to eliminate repeated torch.cat in TokenToKVPoolAllocator

### PR DESCRIPTION
**Pull request by ccyrene**
## Problem

`TokenToKVPoolAllocator.free()` calls `torch.cat` on **every invocation**, creating a new GPU tensor and copying the entire `free_pages` pool each time. When N requests finish in the same scheduling cycle, this triggers N `torch.cat` calls with quadratically growing copy sizes.

## Fix

Accumulate freed tensors in a Python list (`_pending_free`) and flush with a **single `torch.cat`** only when `alloc()` is called. `available_size()` tracks pending slots so the scheduler always sees accurate free memory.

## Benchmark (CUDA, BS=32, 256 tokens/req, pool=65536)

| | Median | p95 |
|---|---|---|
| Before (N torch.cat per cycle) | 527µs | 806µs |
| After  (1 torch.cat per cycle) | 103µs | 160µs |
| **Speedup** | **5.14×** | **5.04×** |

## Changes
- `allocator.py`: add `_pending_free` list + flush logic; fix `backup_state()` to flush pending before snapshot
- `test_allocator_lazy_free.py`: 14 unit tests (correctness, available_size, edge cases, free_group interop, need_sort path, backup/restore)

### CODE REVIEW

File: python/sglang/srt/mem_cache/allocator.py [hunk 1]
```diff
@@ -68,6 +68,8 @@ def restore_state(self, state):
         self.free_pages, self.release_pages = state
 
     def backup_state(self):
+        if hasattr(self, "_pending_free"):
+            self._flush_pending_free()
```
[the remark this answers is not in the document]
Comment (ccyrene) [reply python/sglang/srt/mem_cache/allocator.py]: Fixed in latest commit.

File: python/sglang/srt/mem_cache/allocator.py [hunk 2]
```diff
@@ -136,12 +138,31 @@ def clear(self):
...
+        self._pending_free = []
+        self._pending_free_count = 0
+        if self.need_sort:
+            self.release_pages = torch.cat((self.release_pages, batch))
+        else:
+            self.free_pages = torch.cat((self.free_pages, batch))
 
     def alloc(self, need_size: int):
+        self._flush_pending_free()
+
```
Comment (cs-cat) [inline python/sglang/srt/mem_cache/allocator.py]: Could `_flush_pending_free` be delayed until memory pressure occurs, specifically when `free_pages` tensors are insufficient but `pending_free` + `len(free_pages)` is sufficient, before finally flushing and freeing? This could further reduce temporary tensor allocations in the `alloc`/`free` paths, and would also help with requests with small `bs` values, even 1.

This strategy should be safe when `need_sort=False`.

Comment (ccyrene) [reply python/sglang/srt/mem_cache/allocator.py]: Great point. Would you prefer this in the current PR, or as a follow-up?

Comment (cs-cat) [reply python/sglang/srt/mem_cache/allocator.py]: I suggest doing all the optimizations in a single PR. Thank you for your work!

Comment (ccyrene) [reply python/sglang/srt/mem_cache/allocator.py]: Done — flush now only triggers when free_pages is insufficient, covering both need_sort=False and need_sort=True paths.

File: python/sglang/srt/mem_cache/allocator.py [hunk 3]
```diff
@@ -136,14 +138,33 @@ def clear(self):
...
+            self._pending_free[0]
+            if len(self._pending_free) == 1
+            else torch.cat(self._pending_free)
+        )
+        self._pending_free = []
+        self._pending_free_count = 0
+        if self.need_sort:
+            self.release_pages = torch.cat((self.release_pages, batch))
+        else:
+            self.free_pages = torch.cat((self.free_pages, batch))
```
Comment (cs-cat) [inline python/sglang/srt/mem_cache/allocator.py:161]: why invoke `torch.cat` twice?

**Review by cs-cat — COMMENTED** [review]
btw, if you want your PR to be merged, it's best to ping the code owners to request their reviews.
